### PR TITLE
Suppress duplicate Visual Studio tooltips

### LINKED ISSUE #11611: Visual Studio: IntelliSense tooltips reported twice
### Steps to reproduce:
1. Open Visual Studio 2019
2. Open a C# project
3. Enable reporting of Tooltips
4. Trigger an IntelliSense autocomplete suggestion by typing something.
5. Arrow through the suggestions

### Actual behavior:
The selected item is announced, followed by twice the tooltip.

### Expected behavior:
The selected item is announced, followed by once the tooltip.

#### NVDA installed/portable/running from source:
Installed

#### NVDA version:
alpha-20957

#### Windows version:
Windows 10 2004

#### Name and version of other software in use when reproducing the issue:
Visual Studio 2019 16.7.3 Enterprise

#### Does the issue still occur after restarting your computer?
Yes

#### Have you tried any other versions of NVDA? If so, please report their behaviors.
No

#### If addons are disabled, is your problem still occuring?
Yes

#### Did you try to run the COM registry fixing tool in NVDA menu / tools?
Yes

### PR DESCRIPTION
**Pull request by LeonarddeR**
### Link to issue number:
Closes #11611 

### Summary of the issue:
In Visual Studio, the tooltip opened event is fired twice for IntelliSense popups.

### Description of how this pull request fixes the issue:
In an overlay class, ignore the event if it was fired less than 0.2 seconds ago and had the same name.

### Testing performed:
Tested that tool tips are only announced once.

### Known issues with pull request:
None
### Change log entry:
* Bug fixes
    + In Visual Studio, IntelliSense tool tips that provide additional details about the currently selected IntelliSense item are now only reported once. (#11611)

### CODE REVIEW

**Comment by feerrenrut** [comment]
Can you expand on how you chose 0.2 seconds?

File: source/NVDAObjects/UIA/VisualStudio.py [hunk 1]
```diff
@@ -53,6 +54,22 @@ class IntelliSenseLiveRegion(UIA):
...
+	"""
+
+	_lastToolTipOpenedInfo = (None, None)  #: Keeps track of the last ToolTipOpened event (text, time)
+
+	def event_UIA_toolTipOpened(self):
+		oldText, oldTime = self._lastToolTipOpenedInfo
+		newText = self.name
+		newTime = time.time()
+		self.__class__._lastToolTipOpenedInfo = (newText, newTime)
+		if newText == oldText and oldTime is not None and (newTime - oldTime) < 0.2:
```
Comment (feerrenrut) [inline source/NVDAObjects/UIA/VisualStudio.py:68]: I think this really needs a comment to explain why this is done.

For instance, it is probably worth pointing out that a user may intentionally be trying to hear the same tooltip twice, hence the 0.2 seconds. I think the value could be larger (eg 1 second). Given that the text must be equal consecutively, we should look to approach the limit of how rapidly a user could ever intentionally try to trigger the same tooltip twice in a row.

**Review by feerrenrut — CHANGES_REQUESTED** [review]

**Comment by leonardder** [comment]
> Can you expand on how you chose 0.2 seconds?

This was by trial and error, I was looking for a small window. I can give a more in depth rationalisation based on logging if you desire.

> I think the value could be larger (eg 1 second).

I was just looking for the smallest window. This was actually mostly copied from the should accept UIA live region code, however there, the interval was 0.5 seconds. I could think of situations where one would rapidly arrow back and forth between IntelliSense items, so my feeling says one second might be too much.

**Comment by feerrenrut** [comment]
> I could think of situations where one would rapidly arrow back and forth between IntelliSense items

In this case the `newText == oldText` test should fail though, right?

**Comment by leonardder** [comment]
> 
> 
> > I could think of situations where one would rapidly arrow back and forth between IntelliSense items
> 
> In this case the `newText == oldText` test should fail though, right?

Not in the following situation:

1. Item a has focus, tool tip is shown
2. User presses down arrow, item b is focused but tool tip takes a little while to come up
3. Before the tool tip comes up, user presses up arrow. Item a gets focus again, but tool tip is ignored if the time range is too long.

**Comment by feerrenrut** [comment]
Ok, yes that situation needs to be considered too. 

Could we tie this to input somehow? EG make this longer timeout, but reset when there is keyboard input?

**Comment by leonardder** [comment]
> Could we tie this to input somehow? EG make this longer timeout, but reset when there is keyboard input?

I'm afraid that's pretty hacky. Not sure whether to register the reset, then.

**Comment by feerrenrut** [comment]
I don't think there is any other reasonable heuristic we can use to determine whether the tooltip showing is the user's intention. But basing this solely on keyboard input would exclude low-vision mouse users.

I think for now we can accept this PR the way it is, hopefully this workaround improves the outcome most of the time. For users on slower computers (or under-load or with different tooltip settings) the timeout may not be appropriate. Perhaps one step better is to make the timeout an advanced configuration.

There is a certain argument that problems like this should be left exposed so they are more obvious and get fixed in the 3rd party application (Visual Studio in this case), however we should be able to replicate this issue with accEvent and request a fix from Microsoft.

**Comment by leonardder** [comment]
> There is a certain argument that problems like this should be left exposed so they are more obvious and get fixed in the 3rd party application (Visual Studio in this case), 

Agreed, but it is pretty irritating for daily users.

> however we should be able to replicate this issue with accEvent and request a fix from Microsoft.

I will try to contact Dante Gange about this.

File: source/NVDAObjects/UIA/VisualStudio.py [hunk 2]
```diff
@@ -53,6 +54,22 @@ class IntelliSenseLiveRegion(UIA):
...
+			return
+		super().event_UIA_toolTipOpened()
```
Comment (feerrenrut) [inline source/NVDAObjects/UIA/VisualStudio.py:70]: I'd like to clarify this block.
```suggestion
	_lastToolTipOpenedInfo = (None, None)  #: Keeps track of the last ToolTipOpened event (text, time)
	_preventDuplicateToolTipSeconds = 0.2 #: The duplicate tooltip events will be dropped within this time window

	def event_UIA_toolTipOpened(self):
		oldText, oldTime = self._lastToolTipOpenedInfo
		newText = self.name
		newTime = time.time()
		self.__class__._lastToolTipOpenedInfo = (newText, newTime)
		withinPossibleDupToolTipTimeWindow = (
			oldTime is not None
			and (newTime - oldTime) < _preventDuplicateToolTipSeconds
		)
		if newText == oldText and withinPossibleDupToolTipTimeWindow:
			# Tool-tip event suspected to be a duplicate, drop the event.
			# - Users attempting to rapidly re-announce tool-tips may
			#   have the announcement erroneously  suppressed
			# - Users on slower systems (or systems under load) may still
			#   receive duplicate announcements
			return
		super().event_UIA_toolTipOpened()
```

**Review by feerrenrut — CHANGES_REQUESTED** [review]

**Review by feerrenrut — DISMISSED** [review]
Thanks @leonardder

**Review by feerrenrut — DISMISSED** [review]

**Review by feerrenrut — APPROVED** [review]

**Pull request merged**